tools/mason/MasonArgParse.chpl

```diff
@@ -0,0 +1,236 @@
+
+// A generic argument parser error
+  class ArgumentError : Error {
+    var msg:string;
+    proc init(msg:string) {
+      this.msg = msg;
+    }
+    override proc message() {
+      return msg;
+    }
+  }
+  
+// indicates a result of argument parsing
+  class Argument {
+    var present: bool=false;
+    var values: list(string);     
+    
+    proc getValue(){     
+      return this.values.first();
+    }
+    iter getValues(){
+      for val in values {
+        yield val;
+      }      
+    }    
+    proc hasValue(){
+      return !this.values.isEmpty();
+    }
+  }
+
+  // stores the definition of an option
+  class Action {
+    var name:string;
+    var numOpts:int;
+    var opts:[0..numOpts-1] string;
+    var numArgs:range;
+
+    // TODO: Decouple the argument from the action
+    // maybe pass a list to fill by reference and have the argparser populate
+    // the argument instead?
+    // also need a bool by ref to indicate presence of arg or not
+    proc match(args:[?argsD]string, startPos:int, ref myArg:Argument) throws {
+      var high = 0;
+      
+      // TODO: Replace this high bound with something more reasonable
+      if !this.numArgs.hasHighBound() {
+        high = 10000000000;
+      } else {
+        high = this.numArgs.high;
+      }
+      writeErr("expecting between " + numArgs.low:string + " and "+high:string);
+      var matched = 0;
+      var pos = startPos;
+      var next = pos+1;
+      writeErr("starting at pos: " + pos:string);
+      while matched < high && next <= argsD.high && !args[next].startsWith("-"){
+        pos=next;
+        next+=1;
+        matched+=1;
+        myArg.values.append(args[pos]);
+        myArg.present=true;
+        writeErr("matched val: " + args[pos] + " at pos: " + pos:string);     
+      }
+      if matched < this.numArgs.low {
+        throw new ArgumentError("\\".join(opts) + " not enough values");
+      }
+      return next;
+    }
+ }
+
+  record argumentParser {
+    var result: map(string, shared Argument);
+    var actions: map(string, owned Action);
+    var options: map(string, string);
+
+    proc parseArgs(arguments:[?argsD]string) throws {
+      compilerAssert(argsD.rank==1, "parseArgs requires 1D array");
+      writeErr("start parsing args");   
+      var k = 0;
+      // identify optionIndices where opts start
+      var optionIndices : map(string, int);
+      const zArgsD = {0..argsD.size-1};
+      var zeroArgs = arguments.reindex(zArgsD);
+      var argsList = new list(zeroArgs);
+      
+      for i in zArgsD {
+        // look for = sign after opt, split into two elements
+        if zeroArgs[i].startsWith("-") && zeroArgs[i].find("=") > 0 {
+          var elems = new list(zeroArgs[i].split("=", 1));
+          // replace this opt=val with opt val
+          argsList.pop(i);
+          argsList.insert(i, elems.toArray());
+        }
+      }
+      
+      for i in argsList.indices {        
+        if options.contains(argsList.this(i)) {
```

Comment:
You can write `argsList[i]` instead of `argList.this(i)`.

Also, `argsList.this(i)` aka `argsList[i]` is used in several places here. Consider writing

`const ref listElt = argsList[i];` and then using `listElt` instead.